File: PixelSolution/PixelEngine/Animation2D.cpp
#include "pch.h"
#include "Animation2D.h"
#include "Renderer2D.h"
#include "Registry.h"
#include "PixelEngineAPI.h"
#include "PixelEngine.h"
#include "ResourceManager.h"
#include "Graphics.h"

extern PixelEngine* Engine;
Animation2DData* Animation2D_Add(unsigned int id)
{
	auto registry = GetRegistry();
	registry->Add<Animation2DData>(id);
	registry->Add<Animation2DDList>(id);
	auto data = registry->Get<Animation2DData>(id);
	data->thisID = id;
	return data;
}

Animation2DData* Animation2D_Get(unsigned int id)
{
	auto registry = GetRegistry();
	return registry->Get<Animation2DData>(id);
}

bool Animation2D_Has(unsigned int id)
{
	auto registry = GetRegistry();
	return registry->Has<Animation2DData>(id);
}

void Animation2D_Play(unsigned int id, int AnimationIndex)
{
	auto registry = GetRegistry();
	auto data = registry->Get<Animation2DDList>(id);
	if (data != nullptr)
	{
		if (data->selectIndex != AnimationIndex)
		{
			data->selectIndex = AnimationIndex;
			data->play = true;
			data->selectAnimation = data->animationArray[AnimationIndex];

			auto render = registry->Get<GraphicsData>(id);
			if (render != nullptr)
			{
				render->renderingData.sprite.isShared = false;
			}
			else
			{
				PixelLog::Error("[Animation2D][Play] Not Find Renderer2D");
			}
		}
	}
	else
	{
		PixelLog::Error("[Animation2D][Play] Not Find");
	}
}

int Animation2D_GetFrameIndex(unsigned int id, int AnimationIndex)
{
	auto registry = GetRegistry();
	auto data = registry->Get<Animation2DDList>(id);
	if (data == nullptr)
	{
		PixelLog::Error("[Animation2D][GetFrameIndex] Not Find");
		return -1;
	}

	if (AnimationIndex < 0 || AnimationIndex >= static_cast<int>(data->animationArray.size()))
	{
		PixelLog::Error("[Animation2D][GetFrameIndex] Invalid AnimationIndex");
		return -1;
	}

	if (data->selectIndex == AnimationIndex)
	{
		return data->selectAnimation.framesIndex;
	}

	return data->animationArray[AnimationIndex].framesIndex;
}

void Animation2D_SetFrameIndex(unsigned int id, int AnimationIndex, int FrameIndex)
{
	auto registry = GetRegistry();
	auto data = registry->Get<Animation2DDList>(id);
	if (data == nullptr)
	{
		PixelLog::Error("[Animation2D][SetFrameIndex] Not Find");
		return;
	}

	if (AnimationIndex < 0 || AnimationIndex >= static_cast<int>(data->animationArray.size()))
	{
		PixelLog::Error("[Animation2D][SetFrameIndex] Invalid AnimationIndex");
		return;
	}

	auto& animation = data->animationArray[AnimationIndex];
	const int totalFrames = animation.maxFramesX * animation.maxFramesY;
	if (FrameIndex < 0 || FrameIndex >= totalFrames)
	{
		PixelLog::Error("[Animation2D][SetFrameIndex] Invalid FrameIndex");
		return;
	}

	animation.framesIndex = FrameIndex;
	animation.nowFrameTime = 0.0f;

	if (data->selectIndex == AnimationIndex)
	{
		data->selectAnimation.framesIndex = FrameIndex;
		data->selectAnimation.nowFrameTime = 0.0f;
	}
}

void Animation2D_SetUseUnscaledTime(unsigned int id, bool UseUnscaledTime)
{
	auto registry = GetRegistry();
	auto data = registry->Get<Animation2DDList>(id);
	if (data == nullptr)
	{
		PixelLog::Error("[Animation2D][SetUseUnscaledTime] Not Find");
		return;
	}

	data->useUnscaledTime = UseUnscaledTime;
}

void Animation2D_Stop(unsigned int id)
{
	auto registry = GetRegistry();
	auto data = registry->Get<Animation2DDList>(id);
	if (data != nullptr)
	{
		data->play = false;
	}
	else
	{
		PixelLog::Error("[Animation2D][Stop] Not Find");
	}
}

void Animation2D_Reset(unsigned int id)
{
	auto registry = GetRegistry();
	auto data = registry->Get<Animation2DDList>(id);
	if (data != nullptr)
	{
		data->selectAnimation.framesIndex = 0;
		data->selectAnimation.nowFrameTime = 0.0f;
	}
	else
	{
		PixelLog::Error("[Animation2D][Reset] Not Find");
	}
}

int Animation2D_Create(unsigned int id, const char* textureName, int MaxFramesX, int MaxFramesY, float speed)
{
	auto registry = GetRegistry();
	auto data = registry->Get<Animation2DDList>(id);
	if (data != nullptr)
	{
		int index = data->animationArray.size();
		data->animationArray.emplace_back();
		data->animationArray[index].maxFramesX = MaxFramesX;
		data->animationArray[index].maxFramesY = MaxFramesY;
		data->animationArray[index].framesIndex = 0;
		data->animationArray[index].oneFrameTime = 1.0f / MaxFramesX;
		data->animationArray[index].nowFrameTime = 0.0f;
		data->animationArray[index].animationSpeed = speed;
		data->animationArray[index].textureID = Engine->GetResourceID(RESOURCE_TYPE::TEXTURE, textureName);
		data->selectIndex = index;
		data->play = true;
		data->selectAnimation = data->animationArray[index];
		return index;
	}
	return 0;
}

File: Survival/Engine/Animation2D.lua
local ffi = require("ffi")
local dll = ffi.load("PixelEngine")

ffi.cdef[[
	typedef struct {
		unsigned int thisID;
	} Animation2DData;

	Animation2DData* Animation2D_Add(unsigned int ID);
	Animation2DData* Animation2D_Get(unsigned int ID);
	bool Animation2D_Has(unsigned int ID);
	void Animation2D_Play(unsigned int ID,int AnimationIndex);
	void Animation2D_Stop(unsigned int ID);
	void Animation2D_Reset(unsigned int ID);
	int Animation2D_Create(unsigned int ID,const char* TextureName,int MaxFrameX,int MaxFrameY,float Speed);
	int Animation2D_GetFrameIndex(unsigned int ID,int AnimationIndex);
	void Animation2D_SetFrameIndex(unsigned int ID,int AnimationIndex,int FrameIndex);
	void Animation2D_SetUseUnscaledTime(unsigned int ID,bool UseUnscaledTime);
	void Animation2D_SetLoop(unsigned int ID,int AnimationIndex,bool Loop);
	void Animation2D_SetFrameCallback(unsigned int ID,int AnimationIndex,int FrameIndex,bool Repeat);
]]

---@class Animation2DData
local Animation2DData = {}

---@param AnimationIndex number
function Animation2DData:Play(AnimationIndex)
	return dll.Animation2D_Play(self.thisID,AnimationIndex)
end

function Animation2DData:Stop()
	return dll.Animation2D_Stop(self.thisID)
end

function Animation2DData:Reset()
	return dll.Animation2D_Reset(self.thisID)
end

---@param TextureName string
---@param MaxFrameX number
---@param MaxFrameY number
---@param Speed number
---@return number
function Animation2DData:Create(TextureName,MaxFrameX,MaxFrameY,Speed)
	return dll.Animation2D_Create(self.thisID,TextureName,MaxFrameX,MaxFrameY,Speed)
end

---@param AnimationIndex number
---@return number
function Animation2DData:GetFrameIndex(AnimationIndex)
	return dll.Animation2D_GetFrameIndex(self.thisID,AnimationIndex)
end

---@param AnimationIndex number
---@param FrameIndex number
function Animation2DData:SetFrameIndex(AnimationIndex,FrameIndex)
	return dll.Animation2D_SetFrameIndex(self.thisID,AnimationIndex,FrameIndex)
end

---@param UseUnscaledTime boolean
function Animation2DData:SetUseUnscaledTime(UseUnscaledTime)
	return dll.Animation2D_SetUseUnscaledTime(self.thisID,UseUnscaledTime)
end

---@param AnimationIndex number
---@param Loop boolean
function Animation2DData:SetLoop(AnimationIndex,Loop)
	return dll.Animation2D_SetLoop(self.thisID,AnimationIndex,Loop)
end

---@param AnimationIndex number
---@param FrameIndex number
---@param Repeat boolean
function Animation2DData:SetFrameCallback(AnimationIndex,FrameIndex,Repeat)
	return dll.Animation2D_SetFrameCallback(self.thisID,AnimationIndex,FrameIndex,Repeat)
end

local Animation2D_mt = {
	__index = Animation2DData
}

ffi.metatype("Animation2DData", Animation2D_mt)

Animation2D = Animation2D or {}

---@param ID number 
---@return Animation2DData 
function Animation2D.Add(ID) 
	local data = dll.Animation2D_Add(ID)
	data.thisID = ID
	return data 
end 

---@param ID number 
---@return Animation2DData 
function Animation2D.Get(ID) 
	return dll.Animation2D_Get(ID) 
end 

---@param ID number 
---@return boolean 
function Animation2D.Has(ID) 
	return dll.Animation2D_Has(ID) 
end 

return Animation2D